okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java

```diff
@@ -867,6 +864,31 @@ public Builder encodedFragment(String encodedFragment) {
       return this;
     }
 
+    /**
+     * Re-encodes the components of this URL so that it satisfies (obsolete) RFC 2396, which is
+     * particularly strict for certain components.
+     */
+    Builder reencodeForUri() {
+      for (int i = 0, size = encodedPathSegments.size(); i < size; i++) {
+        String pathSegment = encodedPathSegments.get(i);
+        encodedPathSegments.set(i,
+            canonicalize(pathSegment, PATH_SEGMENT_ENCODE_SET_URI, true, false));
+      }
+      if (encodedQueryNamesAndValues != null) {
```

Comment:
yep, this is a possibly-null list that has possibly-null values. that's the best I can do!